node/chainSimulator/chainSimulator.go

```diff
@@ -412,30 +421,119 @@ func (s *simulator) SetStateMultiple(stateSlice []*dtos.AddressState) error {
+		}
+
+		txHashHex, err := s.sendTx(tx)
+		if err != nil {
+			return nil, err
+		}
+
+		transactionStatus = append(transactionStatus, &transactionWithResult{
+			hexHash: txHashHex,
+			tx:      tx,
+		})
+	}
+
 	time.Sleep(delaySendTxs)
 
-	destinationShardID := s.GetNodeHandler(0).GetShardCoordinator().ComputeId(txToSend.RcvAddr)
-	for count := 0; count < maxNumOfBlockToGenerateWhenExecutingTx; count++ {
-		err = s.GenerateBlocks(1)
+	for count := 0; count < maxNumOfBlocksToGenerateWhenExecutingTx; count++ {
+		err := s.GenerateBlocks(1)
 		if err != nil {
 			return nil, err
 		}
 
-		tx, errGet := s.GetNodeHandler(destinationShardID).GetFacadeHandler().GetTransaction(txHashHex, true)
-		if errGet == nil && tx.Status != transaction.TxStatusPending {
-			log.Info("############## transaction was executed ##############", "txHash", txHashHex)
-			return tx, nil
+		txsAreExecuted := s.computeTransactionStatus(transactionStatus)
+		if txsAreExecuted {
+			return getApiTransactionsFromResult(transactionStatus), nil
+		}
+	}
+
+	return nil, errors.New("something went wrong. Transaction(s) is/are still in pending")
+}
+
+func (s *simulator) computeTransactionStatus(status []*transactionWithResult) bool {
+	allAreExecuted := true
+	for _, resultTx := range status {
+		if resultTx.result != nil {
+			continue
+		}
+
+		sentTx := resultTx.tx
+		destinationShardID := s.GetNodeHandler(0).GetShardCoordinator().ComputeId(sentTx.RcvAddr)
+		result, errGet := s.GetNodeHandler(destinationShardID).GetFacadeHandler().GetTransaction(resultTx.hexHash, true)
+		if errGet == nil && result.Status != transaction.TxStatusPending {
+			log.Info("############## transaction was executed ##############", "txHash", resultTx.hexHash)
+			resultTx.result = result
+			continue
 		}
+
+		allAreExecuted = false
+	}
+
+	return allAreExecuted
+}
+
+func getApiTransactionsFromResult(txWithResult []*transactionWithResult) []*transaction.ApiTransactionResult {
+	result := make([]*transaction.ApiTransactionResult, 0, len(txWithResult))
+	for _, tx := range txWithResult {
+		result = append(result, tx.result)
 	}
 
-	return nil, errors.New("something went wrong transaction is still in pending")
+	return result
+}
+
+func (s *simulator) sendTx(tx *transaction.Transaction) (string, error) {
+	shardID := s.GetNodeHandler(0).GetShardCoordinator().ComputeId(tx.SndAddr)
+	err := s.GetNodeHandler(shardID).GetFacadeHandler().ValidateTransaction(tx)
+	if err != nil {
+		return "", err
+	}
+
+	node := s.GetNodeHandler(shardID)
+	txHash, err := core.CalculateHash(node.GetCoreComponents().InternalMarshalizer(), node.GetCoreComponents().Hasher(), tx)
+	if err != nil {
+		return "", err
+	}
+
+	txHashHex := hex.EncodeToString(txHash)
+	_, err = node.GetFacadeHandler().SendBulkTransactions([]*transaction.Transaction{tx})
+	if err != nil {
+		return "", err
+	}
+
+	for {
+		txs, _ := node.GetFacadeHandler().GetTransactionsPool("")
```

Comment:
what if the tx is out of the pool before this line? is this possible and should we add an exit case here?

Reply:
let's see, that was the code moved by me, indeed, but not the author. Will come back with an answer and solution to this.

Reply:
clarified and made the code more robust